src/main/java/mate/academy/service/impl/ShoppingCartServiceImpl.java

```diff
@@ -0,0 +1,100 @@
+package mate.academy.service.impl;
+
+import java.util.Optional;
+import lombok.RequiredArgsConstructor;
+import mate.academy.dto.cart.ShoppingCartDto;
+import mate.academy.dto.item.CartItemQuantityRequestDto;
+import mate.academy.dto.item.CartItemRequestDto;
+import mate.academy.exception.EntityNotFoundException;
+import mate.academy.mapper.CartItemMapper;
+import mate.academy.mapper.ShoppingCartMapper;
+import mate.academy.model.Book;
+import mate.academy.model.CartItem;
+import mate.academy.model.ShoppingCart;
+import mate.academy.model.User;
+import mate.academy.repository.book.BookRepository;
+import mate.academy.repository.cart.ShoppingCartRepository;
+import mate.academy.repository.item.CartItemRepository;
+import mate.academy.repository.user.UserRepository;
+import mate.academy.service.ShoppingCartService;
+import org.springframework.security.core.Authentication;
+import org.springframework.stereotype.Service;
+
+@Service
+@RequiredArgsConstructor
+public class ShoppingCartServiceImpl implements ShoppingCartService {
+    private final ShoppingCartRepository shoppingCartRepository;
+    private final ShoppingCartMapper shoppingCartMapper;
+    private final BookRepository bookRepository;
+    private final CartItemRepository cartItemRepository;
+    private final CartItemMapper cartItemMapper;
+    private final UserRepository userRepository;
+
+    @Override
+    public void addItemToCart(Authentication authentication,
+                              CartItemRequestDto cartItemRequestDto) {
+        Book bookFromDb = bookRepository
+                .findById(cartItemRequestDto.getBookId())
+                .orElseThrow(() -> new EntityNotFoundException(
+                "Can not find book with id: "
+                        + cartItemRequestDto.getBookId()));
+
+        User user = getUser(authentication);
+
+        ShoppingCart shoppingCart = shoppingCartRepository.findShoppingCartByUserId(user.getId())
+                .orElseGet(() -> registerNewShoppingCart(user));
+
+        Optional<CartItem> existingCartItem = shoppingCart.getCartItems().stream()
+                .filter(item -> item.getBook().getTitle().equals(bookFromDb.getTitle()))
+                .findFirst();
+        if (existingCartItem.isPresent()) {
+            CartItem cartItemToUpdate = existingCartItem.get();
+            int newQuantity = cartItemRequestDto.getQuantity() + cartItemToUpdate.getQuantity();
+            cartItemToUpdate.setQuantity(newQuantity);
+            cartItemRepository.save(cartItemToUpdate);
+        } else {
+            CartItem cartItem = cartItemMapper.toEntity(cartItemRequestDto);
+            cartItem.setBook(bookFromDb);
+            cartItem.setShoppingCart(shoppingCart);
+            cartItemRepository.save(cartItem);
+        }
```

Comment:
1. I don't like idea to check book by title. There is no guarantee it is unique and may cause problems on duplicates
2. Method needs refactoring, now it's too long and not in logical order
I'd see something like below. Please recheck if working and probably fix some naming
```suggestion
        User user = getUser(authentication);

        ShoppingCart shoppingCart = shoppingCartRepository.findShoppingCartByUserId(user.getId())
                .orElseGet(() -> registerNewShoppingCart(user));

        Optional<CartItem> existingCartItem = shoppingCart.getCartItems().stream()
                .filter(item -> item.getBook().getId().equals(cartItemRequestDto.getBookId()))
                .findFirst();

        CartItem cartItemToUpdate;

        if (existingCartItem.isPresent()) {
            cartItemToUpdate = existingCartItem.get();
            cartItemToUpdate.setQuantity(cartItemRequestDto.getQuantity() + cartItemToUpdate.getQuantity());
        } else {
            cartItemToUpdate = createNewCartItem(cartItemRequestDto, shoppingCart);
        }

        cartItemRepository.save(cartItemToUpdate);
    }

    private CartItem createNewCartItem(CartItemRequestDto cartItemRequestDto, ShoppingCart shoppingCart) {
        Book bookFromDb = bookRepository
                .findById(cartItemRequestDto.getBookId())
                .orElseThrow(() -> new EntityNotFoundException(
                        "Can not find book with id: "
                                + cartItemRequestDto.getBookId()));

        CartItem cartItem = cartItemMapper.toEntity(cartItemRequestDto);
        cartItem.setBook(bookFromDb);
        cartItem.setShoppingCart(shoppingCart);
        return cartItem;
    }
```